### PR TITLE
Add diffusion LoRA request path and worker cache

### LINKED ISSUE #281: [Feature]: LoRA adapter support for vLLM alignment
### 🚀 The feature, motivation and pitch

Follow-up to our previous communication under the [roadmap](https://github.com/vllm-project/vllm-omni/issues/165), I’m opening this issue to propose implementing LoRA support for the vLLM alignment workflow. The multimodal RL projects, e.g., [mm_grpo](https://github.com/leibniz-csi/mm_grpo) are also [looking to adopt vllm-omni as a rollout engine](https://github.com/leibniz-csi/mm_grpo/issues/28). Since MM RL typically fine-tunes only the LoRA adapters (i.e., [FlowGRPO](https://github.com/yifan123/flow_grpo), [DiffusionNFT](https://github.com/NVlabs/DiffusionNFT), this integration would be directly beneficial in the sense that it enables RL training workflows where weights update dynamically.

Having LoRA in vllm-omni also aligns with base vllm design, and it brings additional benefits
- Dynamic adaptation: load/unload adapters without restart
- Memory efficiency: smaller memory footprint than full model copies

Happy to take on the work if no one else is currently assigned.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by dongbo910220**
Following the discussion with @AndyZhou952, this PR implements the initial LoRA support framework. 

## Purpose

Fixes #281

 Add request-level dynamic LoRA support for diffusion models (SD3.5/SDXL). This enables:
  - Load/unload adapters at runtime without service restart
  - Per-worker LRU cache with configurable VRAM budget
  - Path whitelist for security
  
## Test Plan

 1. Start server with SD3.5 and LoRA whitelist:
  python -m vllm_omni.entrypoints.openai.api_server \
      --model stabilityai/stable-diffusion-3.5-large \
      --lora-dirs /path/to/lora-test
  2. Send request with LoRA:
  curl -X POST http://localhost:8000/v1/images/generations \
      -H "Content-Type: application/json" \
      -d '{"prompt": "a boy", "lora": {"name": "rafadan", "local_path": "/path/to/lora.safetensors", "scale": 0.8}}'

## Test Result
- Server started successfully with LoRA whitelist configured
  - Request returned 200, image saved
  - Logs confirm LoRA path validation passed and adapter loaded
  - No OOM or exceptions observed

## Limitations (Future Work)

  - Only supports nn.Linear layers (custom kernels like QKVParallelLinear not patched)
  - Single LoRA per request (multi-LoRA composition not yet supported)
  - No background eviction scheduler (on-demand only)

 Co-authored-by: AndyZhou952

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/data.py [hunk 1]
```diff
@@ -282,6 +282,11 @@ class OmniDiffusionConfig:
     # can restrict layers to adapt, e.g. ["q_proj"]
     # Will adapt only q, k, v, o by default.
     lora_target_modules: list[str] | None = None
+    # Dynamic LoRA serving
+    lora_dirs: list[str] | None = None  # whitelist roots for request-specified LoRA paths
+    max_lora_cache_vram: float = 4.0  # GiB per worker
+    max_lora_cache_cpu: float = 8.0  # GiB per worker (placeholder for future CPU caching)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/data.py]: why and how to tune max_lora_cache_vram and max_lora_cache_cpu?

File: vllm_omni/diffusion/lora/manager.py
Comment (SamitHuang) [inline vllm_omni/diffusion/lora/manager.py:1]: Is it better to re-use (or inherit) the LoRAModelManager, LRUCacheLoRAModelManager, and WorkerLoRAManager in vLLM?

**Comment by SamitHuang** [comment]
1. seems this pr mainly considers dynamic lora loading, whitelist `lora_dirs` in server and loar path in request are required to config. I think static lora loading is also commonly used and should be supported in the first version, where we just need to set `lora_path` in server.

2. please add an example script or the related usgae doc

File: vllm_omni/diffusion/lora/manager.py
Comment (AndyZhou952) [reply vllm_omni/diffusion/lora/manager.py:1]: Based on the current vLLM implementation, defining a separate set of managers is more appropriate since
1. `WorkerLoRAManager` is closely coupled with LLM-specific initialization (embedding/vocab_size), which makes direct reuse less suitable in the diffusion context.
2. If we inherite from the vLLM managers, we will need to override / rewrite the `add_adpter` related LoRA handling logic. Current vLLM returns a boolean while in vLLM-Omni `gpu_worker.py` dict format response `{"status": "error", "error": str(e)}` is expected for rpc.

The current implementation from @dongbo910220 works on linear LoRA. PEFT may need to be incorporated to stay consistent with vLLM and to enable greater flexibility. The PEFT-related logic may also differ from the base vLLM.

File: vllm_omni/diffusion/data.py (hunk 1 above)
Comment (ZJY0516) [reply vllm_omni/diffusion/data.py]: same question. Why we need this

**Review by ZJY0516 — COMMENTED** [review]
what blocks supporting custom kernels like QKVParallelLinear?

**Comment by AndyZhou952** [comment]
> what blocks supporting custom kernels like QKVParallelLinear?

I am working on an implementation integrating peft & vllm lora custom modules (which add QKVParallelLinear support). Will be merged into this work once ready.

File: vllm_omni/diffusion/data.py (hunk 1 above)
Comment (AndyZhou952) [reply vllm_omni/diffusion/data.py]: @dongbo910220 please take a look. vllm appears to use a simple count-based eviction so I think these are not really needed. Same for `lora_evict_interval` below.

**Comment by ZJY0516** [comment]
> > what blocks supporting custom kernels like QKVParallelLinear?
> 
> I am working on an implementation integrating peft & vllm lora custom modules (which add QKVParallelLinear support). Will be merged into this work once ready.

Thanks. If supporting the linear layer within vLLM isn't feasible, then implementing it on the Omni diffusion side would be a significant burden

**Comment by dongbo910220** [comment]
> 1. seems this pr mainly considers dynamic lora loading, whitelist `lora_dirs` in server and loar path in request are required to config. I think static lora loading is also commonly used and should be supported in the first version, where we just need to set `lora_path` in server.
> 2. please add an example script or the related usgae doc 

 Thanks for the suggestion! Will add:
  1. Static LoRA loading: Add --lora-path option for single LoRA at server startup
  2. Example script: Add usage examples for both static and dynamic loading modes

File: vllm_omni/diffusion/data.py (hunk 1 above)
Comment (dongbo910220) [reply vllm_omni/diffusion/data.py]: Done. Switched to count-based LRU to align with vLLM. PTAL.

**Comment by zhtmike** [comment]
I think a `peft`-compatible integration should be considered, including support for loading `PEFT` configs, weights, and related components. So far, the `peft` module has been the most popular way to integrate with LoRA on the Hugging Face Hub, and we should not overlook it.

**Comment by AndyZhou952** [comment]
Please refer to #758 for the PEFT design due to the huge amount of refactoring for PEFT adaptation. Tentatively removed whitelist support to be consistent with the vLLM behavior.
